PR merge improvements: auto-merge, edit commit body

mislav opened the pull request:
- Adds `pr merge --auto`, `pr merge --disable-auto` modes
- Folds in https://github.com/cli/cli/pull/2896

```
$ gh pr merge --auto --squash 40
✓ Pull request #40 will automatically get merged via squash when all requirements are met

$ gh pr merge --disable-auto 40
✓ Auto-merge disabled for pull request #40
```

Questions:
- The confirmation checkmark for `merge --auto` is green. Should it be purple?
- The confirmation checkmark for `merge --disable-auto` is green. Should it be red?

Fixes https://github.com/cli/cli/issues/2619
Fixes https://github.com/cli/cli/issues/1023
Reverts #1627, reasoning https://github.com/cli/cli/issues/2954#issuecomment-779390768, ref. https://github.com/cli/cli/issues/1621

TODO:
- [x] tests for `--auto`
- [x] tests for `--disable-auto`
- [x] GHE compatibility for `viewerMergeBodyText` GraphQL field

cristiand391 committed:
Allow editing commit msg when squash merging a PR

samcoe committed:
Fix tests and polish

cristiand391 committed:
Pre-populate default merge commit message if no body was provided

mislav committed:
Add `pr merge --auto`

mislav committed:
Let the server choose the commit subject for squashed merge

For single-commit PRs, the commit subject will be the subject of the
head commit and the PR number. For multi-commit PRs, the commit subject
will be the PR title and PR number. Instead of trying to replicate this
logic client-side, omit the `commitHeadline` param and let the server
apply defaults appropriately.

Reverts https://github.com/cli/cli/pull/1627

mislav committed:
Fix typo

ampinsk commented on pkg/cmd/pr/merge/merge.go:
```diff
@@ -148,51 +162,100 @@ func mergeRun(opts *MergeOptions) error {
+				case PullRequestMergeMethodSquash:
+					method = " via squash"
+				}
+				fmt.Fprintf(opts.IO.ErrOut, "%s Pull request #%d will automatically get merged%s when all requirements are met\n", cs.SuccessIconWithColor(cs.Green), pr.Number, method)
```
Just a small wording tweak
```suggestion
				fmt.Fprintf(opts.IO.ErrOut, "%s Pull request #%d will be automatically merged%s when all requirements are met\n", cs.SuccessIconWithColor(cs.Green), pr.Number, method)
```

I don't love "when all requirements are met" but I can't think of anything better either 🤔

samcoe replied:
Just curious if displaying the merge methods here is accurate? From the API request for enabling automerge it doesn't seem like we specify how the PR is to be merged.

ampinsk reviewed (commented):
Overall looks good to me!

> The confirmation checkmark for merge --auto is green. Should it be purple?
> The confirmation checkmark for merge --disable-auto is green. Should it be red?

I actually think it makes sense for these to be green, since neither of these result in something merging or closing right then and there. 👍

samcoe commented on pkg/cmd/pr/merge/http.go:
```diff
@@ -0,0 +1,100 @@
+package merge
```
I like the extraction of the API code into this new file.

samcoe commented on pkg/cmd/pr/merge/merge.go:
```diff
@@ -131,6 +132,19 @@ func mergeRun(opts *MergeOptions) error {
 
+	isTerminal := opts.IO.IsStdoutTTY()
+
+	if opts.AutoMergeDisable {
```
It seems like enabling automerge should also short circuit the flow of the command. What was behind the decision to move that functionality down into the `mergePullRequest` function?

mislav replied:
@samcoe Since enabling auto-merges accepts identical inputs as the `mergePullRequest` mutation, I've decided to keep all behaviors of normal `pr merge` intact in `--auto` mode, including the interactive ones.

samcoe commented on pkg/cmd/pr/merge/merge.go:
```diff
@@ -29,9 +30,12 @@ type MergeOptions struct {
+	MergeMethod  PullRequestMergeMethod
 
-	Body *string
+	AutoMerge        bool
```
What do you think about changing this to `AutoMergeEnable`?

mislav committed:
Update wording for auto-merge confirmation

Co-authored-by: Amanda Pinsker <ampinsk@github.com>

mislav committed:
Separately query `viewerMergeBodyText` for GHE compatibility

GHE versions 2.22 and older will not have this GraphQL field. Avoid the
resulting error and have the command proceeed with empty text as the
default.

mislav committed:
Allow `pr merge --body ''` to prevent having the default body applied

mislav committed:
Add tests for `pr merge --auto/--disable-auto`

mislav committed:
Merge remote-tracking branch 'origin' into auto-merge

samcoe reviewed (approved):
This looks great, nice work!
